### PR TITLE
Add avl tree implementation

### PR DESCRIPTION
**Pull request by febo**
PR to add an AVL tree implementation using the `NodeAllocator`. AVL trees (theoretically) offer faster lookups as trees are more strictly balanced compared to red-black ones – although they can be slower at insert/remove operations given that they require more rotations to maintain the tree balanced.

The tests included are the same ones as the red-black tree. Adding lookup tests for both avl and red-black trees can illustrate the difference in performance:
```
for _ in 0..200 {
    for k in keys.iter() {
        assert!(tree.get(k) != None);
    }
}
```

### CODE REVIEW

**Comment by jarry-xiao** [comment]
This is super cool! Will take a look later today

File: avl-tree/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,589 @@
+use bytemuck::{Pod, Zeroable};
+use node_allocator::{NodeAllocator, ZeroCopy, SENTINEL};
+use std::{
+    cmp::max,
+    ops::{Index, IndexMut},
+};
+
+// The number of registers.
+const REGISTERS: usize = 3;
```
Comment (jarry-xiao) [inline avl-tree/src/lib.rs:9]: I think there probably needs to be 4 registers in order for bytemuck to properly align the bytes on 64-bit systems

Comment (febo) [reply avl-tree/src/lib.rs:9]: Right, I will change that. I will refactor the code to use this "extra" register to store the parent reference and simplify the logic to traverse the tree.

Comment (febo) [reply avl-tree/src/lib.rs]: I increased the number of registers and considered storing the parent reference in it. Found that there is a trade-off between adding the logic to store/update the parent reference and the current logic to traverse the tree without it. In the end, I am not sure whether there is a significant difference/gain or not, so decided not to change it.

Comment (jarry-xiao) [reply avl-tree/src/lib.rs]: Storing the parent would be a nice touch, but I don't think it's strictly needed for this implementation. Feel free to leave it as a TODO. Awesome work!

File: avl-tree/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+        self.set_field(left, Field::Right, index);
+
+        left
+    }
+
+    fn update_height(&mut self, index: u32) {
+        let left = self.get_field(index, Field::Left);
+        let right = self.get_field(index, Field::Right);
+
+        let height = if (left + right) == 0 {
```
Comment (jarry-xiao) [inline avl-tree/src/lib.rs:337]: Nit: I think this check makes more sense as `left == SENTINEL && right == SENTINEL` even though this is more verbose

File: avl-tree/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+
+        left_height - right_height
+    }
+
+    fn left_rotate(&mut self, index: u32) -> u32 {
+        let right = self.get_field(index, Field::Right);
+        let right_left = self.get_field(right, Field::Left);
+
+        self.set_field(index, Field::Right, right_left);
+        self.set_field(right, Field::Left, index);
```
Comment (jarry-xiao) [inline avl-tree/src/lib.rs:318]: This is a question more for my understanding: why are the respective heights of each node not updated in this function? At first glance it looks like the pointers get misaligned, but I'm also not sure what you do after calling these helpers:


```
      I
         \
           R
          /
       RL
```
becomes

```
          R
        /
      I
         \
           RL
```

But what happens to the previous parent of `index`? There's still going to be a dangling pointer, unless this is handled elsewhere. Additionally, it seems like the heights of `index` and `right` should be swapped

File: avl-tree/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+            let balance_factor = self.balance_factor(left, right);
+
+            let index = if balance_factor > 1 {
+                let left_left = self.get_field(left, Field::Left);
+                let left_right = self.get_field(left, Field::Right);
+                let left_balance_factor = self.balance_factor(left_left, left_right);
+
+                if left_balance_factor < 0 {
+                    let index = self.left_rotate(left);
+                    self.set_field(*child, Field::Left, index);
```
Comment (jarry-xiao) [inline avl-tree/src/lib.rs:378]: I see, this is where you correct the pointers from the above question

File: avl-tree/src/lib.rs (hunk 3 above)
Comment (jarry-xiao) [reply avl-tree/src/lib.rs:318]: Additionally, I don't understand how this decreases the height of the tree? Not well versed on AVL, but my general understanding of rotations is that they should reduce the overall tree height in certain situations.

**Review by jarry-xiao — CHANGES_REQUESTED** [review]
Overall, the implementation looks clean. I need to look into the AVL tree algorithm more to understand the exact details here, but I'll approve these changes after you respond to the comments

File: avl-tree/src/lib.rs (hunk 1 above)
Comment (febo) [reply avl-tree/src/lib.rs]: Thanks! I added a TODO note about the possibility of using the extra register to store the parent reference.

File: avl-tree/src/lib.rs (hunk 2 above)
Comment (febo) [reply avl-tree/src/lib.rs]: Done – now it is clearer the purpose of the comparison.

File: avl-tree/src/lib.rs (hunk 3 above)
Comment (febo) [reply avl-tree/src/lib.rs:325]: The `*-rotate` functions do not update the parent of the `index` node as the parent is not passed as a parameter. They return the reference to the top-most node after the rotation, which then is used to update the parent. In most cases, a rotation decreases the height of the tree, but some rotations are used in preparation to another one and might not decrease the height at first – e.g., left-right rotation:
```
    Z                                Z                                   Y
   / \                              / \                                /   \
  X   d   (left rotation on X) =>  Y   d   (right rotation on Z) =>  X       Z
 / \                              / \                               / \     / \
a   Y                            X   c                             a   b   c   d
   / \                          / \
  b   c                        a   b
```
(`a` `b`, `c` and `d` can be other sub-trees, not necessarily leaf nodes)

The height was updated when the node's balance factor was between `-1` and `1` in the `rebalance` function, which means that there could be cases where the height stored in a register does not precisely reflect the height of the (sub-)tree. I changed the `set_field` function to update the height every time a new child is set. Perhaps there could be edge cases that would require the height of the node to up to date on the `height` register – although I could not come up with one. It does not seem to impact the performance.

File: avl-tree/src/lib.rs (hunk 4 above)
Comment (febo) [reply avl-tree/src/lib.rs:385]: Indeed, the parent is updated outside the `*-rotate` functions.

**Review by jarry-xiao — APPROVED** [review]

**Comment by jarry-xiao** [comment]
Approved! feel free to merge

**Pull request merged**